Refactor agent architecture guides into one

ptgott opened the pull request:
See #37558
Closes #37558 

The Teleport documentation has two guides that explain the architecture of Teleport agents:

- The Database Access Architecture guide (`docs/pages/database-access/architecture.mdx`)
- The SSH Nodes guide (`docs/pages/architecture/nodes.mdx`)

While these guides are framed around describing the architecture of specific agent services, most of the information in these guides applies to Teleport agents in general.

This change refactors these two guides into a single Teleport agent architecture guide that applies to all agent services. Where agents differ, the guide users tables to explain differences. It also links to more specific guides where applicable.

As an alternative, we could create a separate architecture guide for each agent service. This change avoids this approach because:
- The purpose of this change is to refactor existing information for discoverability, and adds minimal new information.
- Agent services differ substantially in how they proxy traffic to specific resources. As of #40115, we expect how-to guides to contain a "How it works" section to explain the architecture of enrolling a specific resource (#39979 includes an example that edits database guides). This approach is cleaner than using a single guide to describe the architecture of connecting to every single resource an agent can proxy.
- The convention of having a separate `[Resource] Access` section of the docs per agent service reflects legacy messaging that we are moving away from.

More specific changes:

- Remove `docs/pages/database-access/architecture`.
- Combine information from the Database Architecture and SSH Nodes guides into an Agent Architecture guide in `docs/pages/architecture/agents.mdx`.
- Move session secording information from the SSH Nodes guide to the Session Recording architecture guide, where it is more appropriate.
- Remove "Cluster State" from the SSH Nodes guide, since this information is present in the backends reference.
- Briefly document direct dial mode for agents, but frame the new guide around the assumption that users will/should enroll agents via the Teleport Proxy Service.
- Add a new architecture diagram and moves the original one to the Database Access Introduction page.

ptgott committed:
Refactor agent architecture guides into one

See #37558

The Teleport documentation has two guides that explain the architecture
of Teleport agents:

- The Database Access Architecture guide
  (`docs/pages/database-access/architecture.mdx`)
- The SSH Nodes guide (`docs/pages/architecture/nodes.mdx`)

While these guides are framed around describing the architecture of
specific agent services, most of the information in these guides applies
to Teleport agents in general.

This change refactors these two guides into a single Teleport agent
architecture guide that applies to all agent services. Where agents
differ, the guide users tables to explain differences. It also links to
more specific guides where applicable.

As an alternative, we could create a separate architecture guide for
each agent service. This change avoids this approach because:
- The purpose of this change is to refactor existing information for
  discoverability, and adds minimal new information.
- Agent services differ substantially in how they proxy traffic to
  specific resources. As of #40115, we expect how-to guides to contain a
  "How it works" section to explain the architecture of enrolling a
  specific resource (#39979 includes an example that edits database
  guides). This approach is cleaner than using a single guide to
  describe the architecture of connecting to every single resource an
  agent can proxy.
- The convention of having a separate `[Resource] Access` section of the
  docs per agent service reflects legacy messaging that we are moving
  away from.

More specific changes:

- Remove `docs/pages/database-access/architecture`.
- Combine information from the Database Architecture and SSH Nodes
  guides into an Agent Architecture guide in
  `docs/pages/architecture/agents.mdx`.
- Move session secording information from the SSH Nodes guide to
  the Session Recording architecture guide, where it is more
  appropriate.
- Remove "Cluster State" from the SSH Nodes guide, since this
  information is present in the backends reference.
- Briefly document direct dial mode for agents, but frame the new guide
  around the assumption that users will/should enroll agents via the
  Teleport Proxy Service.
- Add a new architecture diagram and moves the original one to the
  Database Access Introduction page.

greedy52 commented on docs/pages/desktop-access/introduction.mdx:
```diff
@@ -37,6 +37,12 @@ to Windows computers where you store or manipulate your most sensitive informati
 rather than as a direct replacement for tools that provide general purpose access 
 to Windows computers.
 
+Teleport protects Kubernetes clusters through the Teleport Desktop Service,
```
typo `Kubernetes clusters`?

greedy52 commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,255 @@
+Service](../agents/join-services-to-your-cluster/join-token.mdx#connecting-directly-to-the-auth-service.mdx).
+Once an agent joins a cluster through the Teleport Auth Service, the Teleport
+Proxy Service dials the agent directly, without creating a reverse tunnel. This
+mode supports the Teleport SSH Service, Teleport Desktop Service, and Teleport
+Kubernetes Service.
```
If we were to include discovery service, it doesn't require reverse tunnel either

greedy52 commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,255 @@
+
+|`tsh` command|Upstream infrastructure resource|
+|---|---|
+|`tsh proxy app`|[TCP applications](../application-access/guides/tcp.mdx)|
```
This works for HTTP apps as well

greedy52 commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,255 @@
+
+### Communication between agents and infrastructure resources
+
+Teleport agents proxy user traffic to resources in your infrastructure. The
+method that agents use to connect to a resource typically depends on whether the
+resource is self hosted or managed by a cloud provider. 
+
+For self-hosted resources, operators distribute mutual TLS credentials to
+infrastructure resources, including a certificate signed by a Teleport
+certificate authority for validating the resource and a certificate authority
+certificate that the resource can use for validating Teleport-issued client
+certificates. When an agent connects to a resource, it retrieves a signed client
+certificate from the Teleport Auth Service.
+
+For resources managed by cloud providers, Teleport takes advantage of identity
+and access management (IAM) solutions available through the cloud provider. The
+agent accesses credentials to authenticate to the cloud provider, and has
+permissions to perform certain operations against the infrastructure resources
+it protects, e.g., managing a database or assuming another IAM role.
```
These three paragraphs only apply to databases I believe. The mechanism is different for every other type.

ptgott committed:
Respond to greedy52 feedback

- Edit list of database protocols.
- Note additional commands that retreive certificates.
- Add information re: `tsh apps login` to the client cert table.
- Fix Desktop Service introduction mistake.
- Mention the Discovery Service after the list of agent services.
- List the Discovery Service as one that does not require a reverse
  tunnel.
- Mention HTTP apps in the local proxy table.

ptgott commented:
To do:

- [ ] Update information on heartbeats and (possibly) host certs for all services (https://github.com/gravitational/teleport/pull/41009#discussion_r1586346619)

ibeckermayer commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,255 @@
+for the available methods you can use to join an agent to your Teleport cluster.
+
+All agents in a Teleport cluster keep the Auth Service updated on their status
+with periodic ping messages. They report their IP addresses and the values of
+their assigned labels. Clients can access the list of all agent instances in
```
@greedy52 My understanding is that all services connect and stay connected in a similar way. @rosstimothy knows more about the details of this, if I'm not mistaken.

I'm also unclear if this is done on a per agent basis, or per service basis.

ptgott replied:
I've removed this sentence so we can get this merged more quickly. It looks like the heartbeat system has some intricacies that might make more sense to skip for now instead of delve into.

ibeckermayer commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,265 @@
+Auth Service uses the `user` certificate authority to sign the certificate. See
+[Issuing User Certificates](../architecture/authentication.mdx) for more details
+on how it works. If the user authenticates via the Teleport Web UI, the browser
+retrieves a session cookie.
```
I believe this is how it works:
```suggestion
retrieves a session cookie, which corresponds to a user certificate stored by the Proxy Service on behalf of that user.
```

ibeckermayer commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,265 @@
+on how it works. If the user authenticates via the Teleport Web UI, the browser
+retrieves a session cookie.
+
+In most cases, Teleport users must retrieve another user certificate in order to
```
Not sure what "another" means. Another, as in beyond the one retrieved via the `tsh login` command?

ptgott replied:
That's what I meant, e.g., `tsh kube login`. Would that be correct?

ibeckermayer commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,265 @@
+|Kubernetes clusters|The user runs `tsh kube login`, which updates the local kubeconfig to contain a certificate signed by the Teleport user CA. The user can then execute `kubectl` commands against Teleport-protected Kubernetes clusters.|
+|Servers|After running `tsh login`, the user receives a certificate that contains authorization information for servers. The SSH Service checks the certificate when a user authenticates to a Teleport-protected server.|
+|Web applications|The user authenticates to the Teleport Web UI and visits a Teleport-protected application. The Teleport Proxy Service forwards traffic to the application along with Teleport-signed JSON web tokens that the application can verify. For HTTP API applications, TCP applications, and cloud provider APIs, the user runs `tsh apps login` to retrieve a certificate.|
+|Windows desktops|Users authenticate to the Teleport Web UI. The Web UI initiates a WebSocket session with the Teleport Desktop Service and communicates with the service using Teleport Desktop Protocol (TDP). The Desktop Service converts TDP traffic to Remote Desktop Protocol traffic.|
```
```suggestion
|Windows desktops|Users authenticate to the Teleport Web UI. The Web UI initiates a WebSocket session with the Teleport Proxy Service, which forwards traffic through another WebSocket it creates with the Teleport Desktop Service. The Desktop Service forwards RDP traffic to the desired RDP server.|
```

zmb3 replied:
There's only one websocket here (between the browser and the proxy).

The leg between the proxy and the desktop service is just mTLS (optionally inside an SSH tunnel). It is not a websocket.

ibeckermayer commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,265 @@
+|Web applications|The user authenticates to the Teleport Web UI and visits a Teleport-protected application. The Teleport Proxy Service forwards traffic to the application along with Teleport-signed JSON web tokens that the application can verify. For HTTP API applications, TCP applications, and cloud provider APIs, the user runs `tsh apps login` to retrieve a certificate.|
+|Windows desktops|Users authenticate to the Teleport Web UI. The Web UI initiates a WebSocket session with the Teleport Desktop Service and communicates with the service using Teleport Desktop Protocol (TDP). The Desktop Service converts TDP traffic to Remote Desktop Protocol traffic.|
+
+In most cases (except for some self-hosted clusters), the user will have
+connected to the Auth Service through the Teleport Proxy Service. The Auth
+Service forwards user certificates through the Proxy Service over mutual TLS.  
```
```suggestion
In most cases, users will receive certificates from the Auth Service via a connection to the Teleport Proxy Service. The Auth and Proxy Services connect to each other using mutual TLS.
```

ibeckermayer commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,265 @@
+
+## Establishing reverse tunnels
+
+In most cases, when an agent joins a Teleport cluster, it establishes a reverse
+SSH tunnel to the Proxy Service. As such, users do not need to have direct
```
nit: standard nomenclature is "SSH reverse tunnel" rather than "reverse SSH tunnel"
```suggestion
In most cases, when an agent joins a Teleport cluster, it establishes an SSH
reverse tunnel to the Proxy Service. As such, users do not need to have direct
```

ibeckermayer commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,265 @@
+
+![Tunnel Mode](../../img/architecture/k8s-tunnel.png)
+
+The connection between the Proxy Service and a Teleport agent instance is also
+authenticated with mutual TLS. The Proxy Service generates a short-lived X.509
+certificate signed by the cluster's host CA, with the client's identity and
+routing information for an infrastructure resource encoded in it, and uses it to
+authenticate with the Teleport agent.
```
It's authenticated with mTLS, over the SSH reverse tunnel? Or there's a separate connection made? (Why?)

ibeckermayer commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,265 @@
+
+## Connecting to infrastructure through Teleport agents
+
+Teleport users can access resources in their infrastructure with the `tsh`
+clientCLI, the Teleport Web UI, and Teleport Connect. All of these tools route
```
Can't they also access resources with standard clients for some features, e.g. `psql`, `kubectl`?

ibeckermayer commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,265 @@
+method that agents use to connect to a resource typically depends on whether the
+resource is self hosted or managed by a cloud provider. 
+
+For self-hosted resources, operators distribute mutual TLS credentials to
```
For consistency in nomenclature:
```suggestion
For self-hosted resources, administrators must distribute mutual TLS credentials to
```

ibeckermayer commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,265 @@
+resource is self hosted or managed by a cloud provider. 
+
+For self-hosted resources, operators distribute mutual TLS credentials to
+infrastructure resources, including a certificate signed by a Teleport
+certificate authority for validating the resource and a certificate authority
+certificate that the resource can use for validating Teleport-issued client
+certificates. When an agent connects to a resource, it retrieves a signed client
+certificate from the Teleport Auth Service.
```
```suggestion
infrastructure resources, which typically means adding the Teleport CA
to the resource's trust store and giving the resource a Teleport-signed
certificate. The former allows the resource to verify the agent's identity, and
the latter allows the agent to verify the resource's identity.
```

ibeckermayer commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,265 @@
+
+### Communication between clients and agents
+
+Teleport makes several client tools available for access infrastructure
```
```suggestion
Teleport makes several client tools available for accessing infrastructure
```

ibeckermayer commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,265 @@
+- [Teleport Web UI](../connect-your-client/web-ui.mdx)
+
+After retrieving [client credentials](#credentials-for-teleport-clients), these
+tools authenticate to the Teleport Proxy Service and send traffic to Teleport
```
```suggestion
tools are authenticated to the Teleport Proxy Service and can send traffic through it, to Teleport
```

ibeckermayer commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,265 @@
+Depending on the upstream infrastructure resource a user wants to connect to,
+the `tsh` client tool may spin up one or more local proxy servers. These
+authenticate to the Teleport Proxy Service, then create a listener that accepts
+traffic from local client tools. A user can then connect to the local proxy with
```
maybe an example or two to prevent the user from getting lost in the generalities here
```suggestion
traffic from local, non-Teleport client tools (such as `psql` or `kubectl`). A user can then connect to the local proxy with
```

ibeckermayer reviewed (commented):
Good work here, this is a tough project given the diversity of services we support. Definitely helped to solidify my understanding as well as highlight where I'm weak re: Teleport's architecture.

ptgott commented on docs/pages/architecture/agents.mdx:
That's not a bad idea. I'll open an issue!

ptgott committed:
Partially respond to feedback

- Use a clearer first paragraph and replace "proxy" where possible when
  talking about agents.
- Use more accurate column heading in service table.
- Clarify where agents can run and their relationship to services.
- Other accuracy/clarity fixes, including incorporating suggested
  wording.
- Describe example agent-to-resource authentication methods for
  resources besides databases.
- Mention Teleport Connect.
- Remove the mention of agent ping message payloads.

ptgott commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,265 @@
+deploy and manage agents, including on cloud-hosted Teleport Enterprise
+accounts. 
+
+Each agent can connect to multiple infrastructure resources, and a single agent
+can run multiple services, e.g., the Teleport SSH Service and the Teleport
+Kubernetes Service.
```
👍 I'm using "agent" and "service" for consistency with the rest of the docs (until we resolve #41485), but this makes sense to me.

ptgott commented:
Putting this in draft while I address the [meatier comments](https://github.com/gravitational/teleport/pull/41009#issuecomment-2093205175). Thanks for the thorough reviews!

zmb3 commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,265 @@
+
+| Agent | Traffic it proxies |
+|-------|--------------------|
+| SSH Service | SSH traffic to the target SSH Service instance or remote OpenSSH servers.|
```
Your understanding is mostly accurate, though you are SSHing to an OpenSSH service, not to a "terminal"

ptgott committed:
Edit the agent architecture guide

Flatten the heading structure, organizing the guide according to
connections between components, merging the "Retrieving credentials"
sections into sections related to communication between components.

Also clarify the process in which the Proxy Service forwards traffic to
an agent over a reverse tunnel.

zmb3 commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,280 @@
+tocDepth: 3
+---
+
+**Teleport agents** route traffic from users through the Teleport Proxy Service
+to and from resources in your infrastructure. This guide describes the
```
```suggestion
**Teleport agents** route traffic from users through the Teleport Proxy Service
to and from resources in your infrastructure. This guide describes the
```

_Teleport agents route traffic from users through the Teleport Proxy Service_ seems to imply that the traffic path is user-->agent-->proxy. Are you reading it the same way or am I just being pedantic?

ptgott replied:
I've made this, "**Teleport agents** route traffic from users to and from resources in your infrastructure". I think being more general is fine in the intro paragraph, and the guide clarifies the role of the Proxy Service in detail later.

greedy52 commented on docs/pages/architecture/agents.mdx:
```diff
@@ -0,0 +1,280 @@
+- **Teleport agents:** Agents perform authentication against infrastructure
+  resources, route user traffic to those resources, and perform protocol
+  parsing.
+- **Teleport client tools (`tsh`, Teleport Connect, and the Teleport Web UI):**
```
`tbot` ?

ptgott replied:
I haven't mentioned `tbot` at all in this guide, and assume that "client" is the same as "end user". Since this PR has been in the works for a while, I would suggest leaving out `tbot` and adding it in another change if it make sense to!

ptgott committed:
Edit the intro paragraph

ptgott committed:
Fix spelling

ptgott committed:
Fix linter errors

- Update link target paths
- Add the "Cluster state" H2 from the original "Nodes" architecture
  guide to the Backend Reference so the latter doesn't need to link to a
  missing section.
